### PR TITLE
Implement numpy_helper functions directly in cython

### PR DESCRIPTION
**Pull request by jbrockmendel**
cleanup unused bits of src/datetime.pxd

TL;DR this PR implements `get_datetime64_value`, `get_timedelta64_value`, and `get_datetime64_unit` directly in cython instead of getting them via C.  There is no performance penalty.  Weaning off of the src directory will make the build less of a hassle.

AFAICT the main reason why `get_datetime64_unit`, `get_datetime64_value`, and `get_timedelta64_value` (as well as `is_float_object`, is_datetime64_object`...) are implemented in numpy_helper.h and np_datetime.c is because cython/numpy does not make it easy to cimport the relevant symbols: https://github.com/pandas-dev/pandas/blob/master/pandas/_libs/src/numpy.pxd#L481

```
    # Cannot be supported due to ## ## in macro:
    # bint PyArray_IsScalar(object, verbatim work)
```

After some wrangling and [stackoverflowing](https://stackoverflow.com/questions/46015489/using-numpys-pyarray-isscalar-in-cython/46016420#46016420), I found a way to implement these directly in cython.  Importantly, the C code generated is functionally identical to the existing versions, so there shouldn't be a performance hit.

A big part of the diff comes from moving some declarations from np_datetime's pyx file to the pxd file.

### CODE REVIEW

**Comment by gfyoung** [comment]
> so there shouldn't be a performance hit.

Reasonable, but let's back that claim up with `asv` metrics to be sure.

**Comment by jreback** [comment]
instead of this, is there a reason you simply don't move datetime.pxd into np_datetime.pxd? (certainly can do some of the code cleanup, but after)

**Comment by jbrockmendel** [comment]
> Seem reasonable, but let's back that claim up with asv metrics to be sure.

Will do.  Even with affinity my asv results probably merit outside confirmation.

**Comment by jbrockmendel** [comment]
> instead of this, is there a reason you simply don't move datetime.pxd into np_datetime.pxd? (certainly can do some of the code cleanup, but after)

That may be worth doing, but I see it as orthogonal to this.  Cutting numpy_helper out of the equation simplifies things quite a bit (among other things I'm thinking about setup.py where 'include' and 'pxd' specifications seem pretty fragile)

**Comment by jbrockmendel** [comment]
With appropriate skepticism:
```
taskset 5 asv continuous -E virtualenv -f 1.1 master HEAD -b timestamp
[...]
      before           after         ratio
     [d80a4b81]       [ea4bef00]
+     5.95±0.02μs      6.79±0.05μs     1.14  timestamp.TimestampProperties.time_is_leap_year
+     5.71±0.03μs         6.28±1μs     1.10  timestamp.TimestampProperties.time_is_quarter_end
-     7.09±0.02μs      6.17±0.02μs     0.87  timestamp.TimestampProperties.time_is_year_start
-        195±50ns        109±0.2ns     0.56  timestamp.TimestampProperties.time_microsecond
-        862±40ms         159±10ms     0.18  binary_ops.Timeseries.time_timestamp_ops_diff2
-        443±10ms         77.5±4ms     0.18  binary_ops.TimeseriesTZ.time_timestamp_ops_diff2
-        469±30ms         63.1±7ms     0.13  binary_ops.TimeseriesTZ.time_timestamp_ops_diff1
-         155±8ms       19.1±0.2ms     0.12  binary_ops.Timeseries.time_timestamp_ops_diff1

taskset 5 asv continuous -E virtualenv -f 1.1 master HEAD -b timeseries
[...]
       before           after         ratio
     [d80a4b81]       [ea4bef00]
+      17.3±0.9ms         21.6±1ms     1.25  timeseries.TimeSeries.time_add_irregular
+     46.3±0.09μs       54.3±0.2μs     1.17  timeseries.SemiMonthOffset.time_begin_incr
+        132±10ms            150ms     1.13  timeseries.DatetimeIndex.time_dti_tz_factorize
-      20.3±0.1μs      18.4±0.09μs     0.90  timeseries.Offsets.time_custom_bday_incr
-      20.6±0.1μs       17.8±0.1μs     0.86  timeseries.Offsets.time_timeseries_day_apply

```

Digging out the C code produced by cython for the implementation of `get_datetime64_value` here (and stripping out the traceback comments)

```
static CYTHON_INLINE npy_datetime __pyx_f_6pandas_5_libs_6tslibs_11np_datetime_get_datetime64_value(PyObject *__pyx_v_obj) {
  npy_datetime __pyx_r;

  __pyx_r = ((PyDatetimeScalarObject *)__pyx_v_obj)->obval;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L0:;
  return __pyx_r;
}
```

Whereas the version being replaced is:
```
PANDAS_INLINE npy_datetime get_datetime64_value(PyObject* obj) {
        return ((PyDatetimeScalarObject*)obj)->obval;
    }
```

cython is really verbose, but I pinged their mailing list and was told that the verbosity should be stripped out by the compiler.

Same story for `get_datetime64_unit` and `get_timedelta64_value`.

**Comment by jreback** [comment]
@chris-b1 can you review

File: pandas/_libs/tslibs/np_datetime.pxd [hunk 1]
```diff
@@ -5,12 +5,67 @@ from cpython.datetime cimport date, datetime
...
+cdef extern from "numpy/ndarrayobject.h":
+    ctypedef int64_t npy_timedelta
+    ctypedef int64_t npy_datetime
+
+cdef extern from "numpy/ndarraytypes.h":
+    # ctypedef struct npy_datetimestruct:
+    #    int64_t year
+    #    int32_t month, day, hour, min, sec, us, ps, as
+
+    ctypedef enum NPY_DATETIMEUNIT:
```
Comment (chris-b1) [inline pandas/_libs/tslibs/np_datetime.pxd]: Is this enum needed?

File: pandas/_libs/tslibs/np_datetime.pyx [hunk 2]
```diff
@@ -47,6 +27,34 @@ cdef extern from "../src/datetime/np_datetime.h":
     pandas_datetimestruct _NS_MIN_DTS, _NS_MAX_DTS
 
 # ----------------------------------------------------------------------
+# numpy object inspection
+
+cdef inline npy_datetime get_datetime64_value(object obj) nogil:
+    """
```
Comment (chris-b1) [inline pandas/_libs/tslibs/np_datetime.pyx]: Can you make the docstrings more descriptive?  Something like `"returns int64 value underlying numpy datetime scalar Python object"`

**Review by chris-b1 — APPROVED** [review]
Actual changes look fine to me

File: pandas/_libs/tslibs/np_datetime.pxd (hunk 1 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/np_datetime.pxd]: No, it's leftover from an earlier revision.  Will remove.

File: pandas/_libs/tslibs/np_datetime.pyx (hunk 2 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/np_datetime.pyx]: Good idea, will change.

**Comment by jreback** [comment]
rebase as always on re-push

**Comment by jbrockmendel** [comment]
Thoughts on this?  I'd like to follow-up by moving a few more pieces of src/datetime.pxd over to np_datetime.pxd

**Comment by gfyoung** [comment]
ping @jreback

File: pandas/_libs/lib.pyx [hunk 3]
```diff
@@ -48,9 +48,8 @@ from cpython.datetime cimport (PyDateTime_Check, PyDate_Check,
                                PyTime_Check, PyDelta_Check,
                                PyDateTime_IMPORT)
 PyDateTime_IMPORT
-# this is our tseries.pxd
-from datetime cimport get_timedelta64_value, get_datetime64_value
 
+from tslibs.np_datetime cimport get_timedelta64_value, get_datetime64_value
```
Comment (jreback) [inline pandas/_libs/lib.pyx]: so this now makes lib depend on np_datetime.pxd, which is ok as tseries_depends now lists this.

**Comment by jreback** [comment]
thanks! you will prob need to rebase your other PR's on this just to make sure.
